### ISSUE TITLE
Slower Than C 

### ISSUE THREAD

**Issue by arshidkv12** [davidcole1340/ext-php-rs#310]
Hi
I tested with xdebug profile option with grind cache. 
Functions are slower than c. How to speed up?

**Comment by ryangjchandler** [comment]
The difference in speed is likely due to converting between PHP types (`Zval`) and Rust types such as `String`, `Vec`, etc.

**Comment by joehoyle** [comment]
I think to test this assumption it could be a good idea to benchmark:

```
#[php_function]
fn should_be_fast(&Zval zval ) -> Zval {
    zval.shallow_clone()
}
```

I suspect we might actually have several inefficiencies in the calling bridging, unnecessarily allocating etc.

**Comment by ryangjchandler** [comment]
That would just be benchmarking a direct clone of a Zval though right? Not the `Vec<T>` allocations, etc?

**Comment by joehoyle** [comment]
I'm not sure, I think the `#[php_function]` may generate code that has additional allocations etc in.

**Comment by EdmondDantes** [comment]
The current approach is expected to introduce some performance overhead:

Calling a function from a wrapper function.
Copying variables to the stack.
In other words, if I create a function whose body does very little useful work and plan to call this function frequently, I will inevitably lose about 50% of performance.

I'm not sure how common this case is, but if necessary, it's possible to create a function without a wrapper and work directly with zend_execute_data* execute_data and zval* return_val.

**Comment by Xenira** [comment]
For calling functions more efficiently i use something like this: https://github.com/Xenira/php-iter-tools/blob/rust-impl/src/lib.rs#L1739

**Comment by arshidkv12** [comment]
I think zig is best for it https://github.com/arshidkv12/zig-php

**Comment by EdmondDantes** [comment]
@arshidkv12 

```zig
export fn hello_world(execute_data: ?*php.zend_execute_data, return_value: ?*php.zval) void {
    _ = execute_data;
    _ = return_value;
    _ = php.php_printf("Hello from ZIG!\n");
}
```

@Xenira 

That's exactly what I'm talking about – the same code can be written in Rust without any performance loss. We need to consider this implementation option.

**Comment by kakserpom** [comment]
@Xenira
Why don't we move [your implementation](https://github.com/Xenira/php-iter-tools/blob/rust-impl/src/lib.rs#L1739) from php-iter-tools into ext-php-rs?

**Comment by Xenira** [comment]
@kakserpom is on my todo list. Will create an issue for this.

**Comment by Xenira** [comment]
Probably want to create performance tests first though

**Comment by Xenira** [comment]
<img width="3425" height="318" alt="Image" src="https://github.com/user-attachments/assets/288f5e3d-6eaf-4c1b-9ff6-07f0d603e57e" />

Speculations are probably correct. Lots of allocations and parsing going on.

**Comment by ptondereau** [comment]
Bringing back to life this issue, before callback optimization, I tried to implement a fast path for non-variadic functions/methods.
Here are the results so far with PHP 8.4.19 NTS, Rust 1.94.1:

| Benchmark | Before (Ir/call) | After (Ir/call) | Speedup |
|-----------|------------------:|----------------:|--------:|
| `bench_function(n)` 1 call | 1,459 | 343 | **4.3x** |
| `bench_function(n)` 100k calls | 1,025 | 343 | **3.0x** |
| `bench_callback(fn, 1)` | 3,736 | 2,386 | **1.6x** |
| `bench_callback(fn, 10)` | 1,491/call | 1,367/call | **1.1x** |
| `bench_callback(fn, 100k)` | 1,241/call | 1,254/call | ~same |

Basically, macros read function arguments directly from PHP's call frame via pointer arithmetic instead of building a heap-allocated argument parser on every call.

**Comment by ptondereau** [comment]
Addressed in #720, here are the callgrind results with PHP 8.4 NTS, Rust stable:

| Benchmark | Before (Ir) | After (Ir) | Speedup |
|-----------|------------:|-----------:|--------:|
| Property read, 1 iteration (4 props) | 5,776 | 1,578 | **3.7x** |
| Property read, 100k iterations | 235,203,424 | 157,800,000 | **1.5x** |
| Property write, 1 iteration (4 props) | 5,556 | 1,336 | **4.2x** |
| Property write, 100k iterations | 220,397,869 | 139,994,445 | **1.6x** |
| LL cache hits (read, 100k) | 219 | 52 | **4.2x fewer** |

Macros now generate static function pointers and compile-time arrays instead of boxing closures and building a HashMap at runtime.
